pkg/reconciler/awssnssource/subscription.go

```diff
@@ -78,46 +79,61 @@ func (r *Reconciler) ensureSubscribed(ctx context.Context) error {
 			"Error creating SNS client: %s", err))
 	}
 
-	resp, err := snsClient.SubscribeWithContext(ctx, &sns.SubscribeInput{
-		Endpoint:              aws.String(url.String()),
-		Protocol:              &url.Scheme,
-		TopicArn:              aws.String(spec.ARN.String()),
-		Attributes:            spec.SubscriptionAttributes,
-		ReturnSubscriptionArn: aws.Bool(true),
-	})
+	topicARN := spec.ARN.String()
 
+	subsARN, err := findSubscription(ctx, snsClient, topicARN, url.String())
 	switch {
-	case isAWSError(err):
-		// All documented API errors require some user intervention and
-		// are not to be retried.
-		// https://docs.aws.amazon.com/sns/latest/api/API_Subscribe.html#API_Subscribe_Errors
-		status.MarkNotSubscribed(v1alpha1.AWSSNSReasonRejected, "Subscription request rejected")
-		return controller.NewPermanentError(susbscribeErrorEvent(url, spec.ARN.String(), err))
+	case isNotFound(err):
+		subsARN, err = subscribe(ctx, snsClient, topicARN, url, spec.SubscriptionAttributes)
```

Comment:
We then use that function in the main reconciliation logic and "gate" the call to `Subscribe()`, which only occurs if no subscription was found for our criteria.